File: ios-objc-sample/NativeDisplaySampleObjc/RootTabBarController.m
#import "RootTabBarController.h"
#import "CleverTapIntegrationViewController.h"
#import "SlotDemoViewController.h"
#import "UIKitDemoViewController.h"
#import "DemoMenuViewController.h"

@implementation RootTabBarController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.viewControllers = @[
        [self makeTab:[CleverTapIntegrationViewController new] title:@"Events" sfSymbol:@"antenna.radiowaves.left.and.right"],
        [self makeTab:[SlotDemoViewController new] title:@"Slots" sfSymbol:@"square.stack.3d.up"],
        [self makeTab:[UIKitDemoViewController new] title:@"UIKit" sfSymbol:@"macwindow"],
        [self makeMoreTab],
    ];
}

- (UIViewController *)makeTab:(UIViewController *)vc title:(NSString *)title sfSymbol:(NSString *)symbol {
    UINavigationController *nav = [[UINavigationController alloc] initWithRootViewController:vc];
    nav.tabBarItem = [[UITabBarItem alloc] initWithTitle:title image:[UIImage systemImageNamed:symbol] tag:0];
    return nav;
}

- (UIViewController *)makeMoreTab {
    DemoMenuViewController *menu = [DemoMenuViewController new];
    UINavigationController *nav = [[UINavigationController alloc] initWithRootViewController:menu];
    nav.tabBarItem = [[UITabBarItem alloc] initWithTitle:@"More" image:[UIImage systemImageNamed:@"ellipsis.circle"] tag:0];
    return nav;
}

@end

File: ios-sample/NativeDisplaySample/UIKit/RootTabBarController.swift
import UIKit
import SwiftUI

/// Programmatic UITabBarController configuring the 5 primary app tabs.
/// UITabBarController retains each UIViewController across tab switches,
/// so state is preserved without any extra work.
class RootTabBarController: UITabBarController {

    override func viewDidLoad() {
        super.viewDidLoad()
        viewControllers = [
            makeTab(
                UIHostingController(rootView: CleverTapIntegrationView()),
                title: "Events",
                sfSymbol: "antenna.radiowaves.left.and.right"
            ),
            makeTab(
                UIHostingController(rootView: SlotDemoView()),
                title: "Slots",
                sfSymbol: "square.stack.3d.up"
            ),
            makeTab(
                UIKitTestViewController(),
                title: "UIKit",
                sfSymbol: "macwindow"
            ),
            makeTab(
                UIKitSlotDemoViewController(),
                title: "UIKit Slots",
                sfSymbol: "square.grid.3x3"
            ),
            makeTab(
                UIHostingController(rootView: TestConfigBrowserView()),
                title: "Browser",
                sfSymbol: "testtube.2"
            ),
            makeTab(
                makeMoreTab(),
                title: "More",
                sfSymbol: "ellipsis.circle"
            ),
        ]
    }

    // MARK: - Private helpers

    private func makeTab(_ vc: UIViewController, title: String, sfSymbol: String) -> UIViewController {
        vc.tabBarItem = UITabBarItem(
            title: title,
            image: UIImage(systemName: sfSymbol),
            tag: 0
        )
        return vc
    }

    private func makeMoreTab() -> UIViewController {
        let hosting = UIHostingController(rootView: NavigationView { MoreMenuView() })
        let nav = UINavigationController(rootViewController: hosting)
        return nav
    }
}
